crates/router/src/core/payments.rs

```diff
@@ -2887,6 +2891,118 @@ where
+    state: &SessionState,
+    merchant_account: &domain::MerchantAccount,
+    key_store: &domain::MerchantKeyStore,
+    connectors: Vec<api::SessionConnectorData>,
+    _operation: &Op,
+    mut payment_data: D,
+    customer: &Option<domain::Customer>,
+    _session_surcharge_details: Option<api::SessionSurchargeDetails>,
+    business_profile: &domain::Profile,
+    header_payload: HeaderPayload,
+) -> RouterResult<D>
+where
+    Op: Debug,
+    F: Send + Clone,
+
+    // To create connector flow specific interface data
+    D: OperationSessionGetters<F> + OperationSessionSetters<F> + Send + Sync + Clone,
+    D: ConstructFlowSpecificData<F, Req, router_types::PaymentsResponseData>,
+    RouterData<F, Req, router_types::PaymentsResponseData>: Feature<F, Req>,
+
+    // To construct connector flow specific api
+    dyn api::Connector:
+        services::api::ConnectorIntegration<F, Req, router_types::PaymentsResponseData>,
+{
+    let call_connectors_start_time = Instant::now();
+    let mut join_handlers = Vec::with_capacity(connectors.len());
+    for session_connector_data in connectors.iter() {
+        let merchant_connector_id = session_connector_data
+            .connector
+            .merchant_connector_id
+            .as_ref()
+            .get_required_value("merchant_connector_id")
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("connector id is not set")?;
+        let merchant_connector_account = state
+            .store
+            .find_merchant_connector_account_by_id(&state.into(), merchant_connector_id, key_store)
+            .await
+            .to_not_found_response(errors::ApiErrorResponse::MerchantConnectorAccountNotFound {
+                id: merchant_connector_id.get_string_repr().to_owned(),
+            })?;
+        let connector_id = session_connector_data.connector.connector.id();
+        let router_data = payment_data
+            .construct_router_data(
+                state,
+                connector_id,
+                merchant_account,
+                key_store,
+                customer,
+                &merchant_connector_account,
+                None,
+                None,
+            )
+            .await?;
+
+        let res = router_data.decide_flows(
+            state,
+            &session_connector_data.connector,
+            CallConnectorAction::Trigger,
+            None,
+            business_profile,
+            header_payload.clone(),
+        );
+
+        join_handlers.push(res);
+    }
+
+    let result = join_all(join_handlers).await;
+
+    for (connector_res, session_connector) in result.into_iter().zip(connectors) {
+        let connector_name = session_connector.connector.connector_name.to_string();
+        match connector_res {
+            Ok(connector_response) => {
+                if let Ok(router_types::PaymentsResponseData::SessionResponse {
```

Comment:
we can instead use a match here right? instead of two if lets

Reply:
No. we are de-structuring `SessionResponse` from `connector_response.response`